Sources/AWSSDKSwiftCore/AWSShapes/Payload.swift

```diff
@@ -16,44 +16,81 @@ import struct Foundation.Data
 import NIO
 import NIOFoundationCompat
 
-/// Object storing request/response payload
-public struct AWSPayload {
+public enum AWSPayload {
 
-    /// Construct a payload from a ByteBuffer
-    public static func byteBuffer(_ byteBuffer: ByteBuffer) -> Self {
-        return Self(byteBuffer: byteBuffer)
-    }
+    case byteBuffer(ByteBuffer)
+    case stream(size: Int, closure: (EventLoop)->EventLoopFuture<ByteBuffer>)
 
     /// Construct a payload from a Data
     public static func data(_ data: Data) -> Self {
         var byteBuffer = ByteBufferAllocator().buffer(capacity: data.count)
         byteBuffer.writeBytes(data)
-        return Self(byteBuffer: byteBuffer)
+        return .byteBuffer(byteBuffer)
     }
 
     /// Construct a payload from a String
     public static func string(_ string: String) -> Self {
         var byteBuffer = ByteBufferAllocator().buffer(capacity: string.utf8.count)
         byteBuffer.writeString(string)
-        return Self(byteBuffer: byteBuffer)
+        return .byteBuffer(byteBuffer)
+    }
+
+    /// Construct a payload from a NIOFileHandle
+    public static func fileHandle(_ fileHandle: NIOFileHandle, size: Int, fileIO: NonBlockingFileIO, byteBufferAllocator: ByteBufferAllocator = ByteBufferAllocator()) -> Self {
+        let blockSize = 64*1024
+        var leftToRead = size
+        func stream(_ eventLoop: EventLoop) -> EventLoopFuture<ByteBuffer> {
+            let blockSize = min(leftToRead, blockSize)
+            leftToRead -= blockSize
+            let futureByteBuffer: EventLoopFuture<ByteBuffer> = fileIO.read(fileHandle: fileHandle, byteCount: blockSize, allocator: byteBufferAllocator, eventLoop: eventLoop)
+                .map { byteBuffer in
+                    precondition(byteBuffer.readableBytes == blockSize, "File did not have enough data")
+                    return byteBuffer
+            }
```

Comment:
If you are talking about the `precondition`. It is something we have to check. It is checking the user has passed us a size that the file is able to deliver. If we ask for 64k, because user tells us there is 64k, and the file is only 32k then we get a bytebuffer of a 32k. AHC is left waiting for the remaing 32k but we keep passing empty bytebuffers to AHC.